scripts/install-mender.sh

```diff
@@ -501,9 +575,9 @@ check_dist_and_version() {
 
 banner
 check_dist_and_version
+get_deps
 init "$@"
 print_components
-get_deps
 add_repo
 do_install_open
 do_install_commercial
```

Comment:
Few things look odd here.

For an start, see what happens on `--help` option:

```
root@0804bc6de38d:/# bash install-mender.sh --help

                          _
 _ __ ___   ___ _ __   __| | ___ _ __
| '_ ` _ \ / _ \ '_ \ / _` |/ _ \ '__|
| | | | | |  __/ | | | (_| |  __/ |
|_| |_| |_|\___|_| |_|\__,_|\___|_|

Running the Mender installation script.
--

  Detected distribution:
	ubuntu/focal
debconf: delaying package configuration, since apt-utils is not installed
(Reading database ... 4469 files and directories currently installed.)
Preparing to unpack .../gpgv_2.2.19-3ubuntu2.2_amd64.deb ...
Unpacking gpgv (2.2.19-3ubuntu2.2) over (2.2.19-3ubuntu2.1) ...
Setting up gpgv (2.2.19-3ubuntu2.2) ...
Selecting previously unselected package readline-common.
(Reading database ... 4469 files and directories currently installed.)
Preparing to unpack .../00-readline-common_8.0-4_all.deb ...
Unpacking readline-common (8.0-4) ...
...
...
Setting up libhx509-5-heimdal:amd64 (7.7.0+dfsg-1ubuntu1) ...
Setting up gpgsm (2.2.19-3ubuntu2.2) ...
Setting up gpg-wks-server (2.2.19-3ubuntu2.2) ...
Setting up libkrb5-26-heimdal:amd64 (7.7.0+dfsg-1ubuntu1) ...
Setting up libheimntlm0-heimdal:amd64 (7.7.0+dfsg-1ubuntu1) ...
Setting up libgssapi3-heimdal:amd64 (7.7.0+dfsg-1ubuntu1) ...
Setting up libldap-2.4-2:amd64 (2.4.49+dfsg-2ubuntu1.9) ...
Setting up dirmngr (2.2.19-3ubuntu2.2) ...
Setting up libcurl4:amd64 (7.68.0-1ubuntu2.13) ...
Setting up curl (7.68.0-1ubuntu2.13) ...
Setting up gpg-wks-client (2.2.19-3ubuntu2.2) ...
Setting up gnupg (2.2.19-3ubuntu2.2) ...
Processing triggers for libc-bin (2.31-0ubuntu9.2) ...
usage: install-mender.sh [options] [component...] [-- [options-for-mender-setup] ]

options: [-h, help] [-c channel] [--demo] [--commercial]
  -h, --help          print this help
  -c CHANNEL          channel: stable(default)|experimental
  --demo              use defaults appropriate for demo
  --commercial        install commercial components, requires --jwt-token
  --jwt-token TOKEN   Hosted Mender JWT token

If no components are specified, defaults will be installed

Anything after a '--' gets passed directly to 'mender setup' command.

Supported components (x = installed by default):
 (x) mender-client
 (x) mender-configure
 (-) mender-configure-demo
 (-) mender-configure-timezone
 (x) mender-connect
 (-) mender-gateway
 (-) mender-monitor
 (-) mender-monitor-demo
```

Obviously, we cannot install nothing on a `--help` call.

Then let's say I want to install one component, i.e. `mender-gateway`:

```
root@8e2c8e45b372:~# bash install-mender.sh --jwt-token $JWT mender-gateway

                          _
 _ __ ___   ___ _ __   __| | ___ _ __
| '_ ` _ \ / _ \ '_ \ / _` |/ _ \ '__|
| | | | | |  __/ | | | (_| |  __/ |
|_| |_| |_|\___|_| |_|\__,_|\___|_|

Running the Mender installation script.
--

  Detected distribution:
	ubuntu/focal
debconf: delaying package configuration, since apt-utils is not installed
(Reading database ... 4469 files and directories currently installed.)
Preparing to unpack .../gpgv_2.2.19-3ubuntu2.2_amd64.deb ...
Unpacking gpgv (2.2.19-3ubuntu2.2) over (2.2.19-3ubuntu2.1) ...
Setting up gpgv (2.2.19-3ubuntu2.2) ...
Selecting previously unselected package readline-common.
...
...
Setting up libcurl4:amd64 (7.68.0-1ubuntu2.13) ...
Setting up curl (7.68.0-1ubuntu2.13) ...
Setting up gpg-wks-client (2.2.19-3ubuntu2.2) ...
Setting up gnupg (2.2.19-3ubuntu2.2) ...
Processing triggers for libc-bin (2.31-0ubuntu9.2) ...
  Detected architecture:
	amd64
  Installing from channel:
	stable
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 51894  100 51894    0     0   194k      0 --:--:-- --:--:-- --:--:--  194k
  Installing commercial components from Mender release: 3.3.0
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 51894  100 51894    0     0   241k      0 --:--:-- --:--:-- --:--:--  241k
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 51894  100 51894    0     0   250k      0 --:--:-- --:--:-- --:--:--  249k
Installing version 1.0.0 of mender-gateway
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 51894  100 51894    0     0  83835      0 --:--:-- --:--:-- --:--:-- 83835
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 51894  100 51894    0     0   272k      0 --:--:-- --:--:-- --:--:--  271k
Installing version 1.2.0 of mender-monitor
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 51894  100 51894    0     0   136k      0 --:--:-- --:--:-- --:--:--  136k
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 51894  100 51894    0     0   103k      0 --:--:-- --:--:-- --:--:--  103k
Installing version 1.2.0 of mender-monitor-demo
  Selected components:
	mender-gateway
OK
Installing Mender APT repository at /etc/apt/sources.list.d/mender.list...
  Installing commercial components from https://downloads.customer.mender.io/content/hosted/
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 51894  100 51894    0     0   171k      0 --:--:-- --:--:-- --:--:--  171k
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 51894  100 51894    0     0   204k      0 --:--:-- --:--:-- --:--:--  204k
Installing mender-gateway (1.0.0)
Selecting previously unselected package mender-gateway.
(Reading database ... 4899 files and directories currently installed.)
Preparing to unpack mender-gateway_1.0.0-1+ubuntu+focal_amd64.deb ...
Unpacking mender-gateway (1.0.0-1+ubuntu+focal) ...
Setting up mender-gateway (1.0.0-1+ubuntu+focal) ...
Package: mender-gateway
Status: install ok installed
Priority: optional
Section: admin
Installed-Size: 5677
Maintainer: Mender Team <mender@northern.tech>
Architecture: amd64
Version: 1.0.0-1+ubuntu+focal
Depends: libc6 (>= 2.4)
Suggests: mender-client
Description: Mender Gateway
 Mender Gateway exposes the Mender Device APIs to local devices
 running the Mender client and forwards the requests to an upstream
 Mender Server. It can optionally enhance the APIs with additional
 edge-specific features.
  Success!
```

See:
* `Installing version 1.2.0 of mender-monitor`, which is not installed
  * This even happens when installing only OS components.
* "Detected architecture:" and "Installing from channel:" in the middle
* stdout and stderror mixed when figuring out the versions

I think we should leave it as it was: banner and all the passive info at the beginning. We don't need to figure out the "latest" versions so early, it is enough to inform that we will install from "stable". Only later we can translate this for the commercial packages into specific versions and print it (and in this way we don't output versions of packages that we are not installing).

Reply:
> stdout and stderror mixed when figuring out the versions

I don't follow. It's only writing to `stderr` on errors. You still want everything written to `stdout`? 

I'm fine with it. Just wondering what you mean

Reply:
Maybe I am wrong and is not `stdout`. I just meant that this part is difficult to read:
```
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 51894  100 51894    0     0   194k      0 --:--:-- --:--:-- --:--:--  194k
  Installing commercial components from Mender release: 3.3.0
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 51894  100 51894    0     0   241k      0 --:--:-- --:--:-- --:--:--  241k
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 51894  100 51894    0     0   250k      0 --:--:-- --:--:-- --:--:--  249k
Installing version 1.0.0 of mender-gateway
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 51894  100 51894    0     0  83835      0 --:--:-- --:--:-- --:--:-- 83835
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 51894  100 51894    0     0   272k      0 --:--:-- --:--:-- --:--:--  271k
Installing version 1.2.0 of mender-monitor
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 51894  100 51894    0     0   136k      0 --:--:-- --:--:-- --:--:--  136k
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 51894  100 51894    0     0   103k      0 --:--:-- --:--:-- --:--:--  103k
Installing version 1.2.0 of mender-monitor-demo
```

But probably is not a problem anymore with your rework. I'll look at it with fresh eyes now :popcorn: